### ISSUE TITLE
`hoistAwait` removes necessary `await` for operators/functions with async autoclosures.

### ISSUE THREAD

**Issue by lhunath** [nicklockwood/SwiftFormat#1531]
The following code should be perfectly fine:
```
// TODO: Remove me when https://github.com/apple/swift/pull/36830 merges.
infix operator ???: NilCoalescingPrecedence
public func ??? <T>(optional: T?, defaultValue: @autoclosure () async throws -> T) async rethrows -> T {
    switch optional {
        case let .some(value):
            return value
        case .none:
            return try await defaultValue()
    }
}
func onLoad() async {
    let organization = await self.sessionInteractor.stage.values.first??.organization ???
        (await self.projectsInteractor.global.organizations().map(\.?.last).removeNil().values.first)
}
```
But it triggers:
```
(hoistAwait) Move inline await keyword(s) to start of expression.
```
The format operation also removes the second `await`. The result of the format operation is:
```
let organization = await self.sessionInteractor.stage.values.first??.organization ???
    (self.projectsInteractor.global.organizations().map(\.?.last).removeNil().values.first)
    // Expression is 'async' but is not marked with 'await'
```

**Comment by nicklockwood** [comment]
Unfortunately SwiftFormat isn't smart enough to recognize that an operator could introduce a new async scope within an expression. I'm not sure what I can do to easily fix this.

If the operator is declared in the same file then SwiftFormat could in theory parse it and handle this case, but in most cases I'd guess the operator would probably be defined elsewhere.

I'll think about it. In the meantime, I suggest either disabling the rule globally for your project, or adding the following comment directive wherever you use this operator:

```swift
// swiftformat:disable:next hoistAwait
```

**Comment by lhunath** [comment]
Thanks. Will this situation change when Swift decides to merge the linked PR and make the default operator `??` support async operations?

**Comment by nicklockwood** [comment]
For now I've added the ability to add operators like `???` to the list of async-capturing functions specified with the `--asynccapturing` option. That will at least can you can work around the issue without need to disable the rule completely. If/when Apple adds `reasync` then I'll add `??` to the list by default.

**Comment by nicklockwood** [comment]
@lhunath support for `--asynccapturing ???` has landed in 0.52.5
